dev/devicelab/lib/tasks/plugin_tests.dart

```diff
@@ -55,30 +63,86 @@ class PluginTest {
   }
 }
 
-class FlutterProject {
-  FlutterProject(this.parent, this.name);
+abstract class _FlutterProject {
+  _FlutterProject(this.parent, this.name);
 
   final Directory parent;
   final String name;
 
-  static Future<FlutterProject> create(Directory directory, List<String> options) async {
+  String get rootPath => path.join(parent.path, name);
+
+  Future<void> delete() async {
+    if (Platform.isWindows) {
+      // A running Gradle daemon might prevent us from deleting the project
+      // folder on Windows.
+      await exec(
+        path.absolute(path.join(rootPath, 'android', 'gradlew.bat')),
+        <String>['--stop'],
+        canFail: true,
+      );
+      // TODO(ianh): Investigating if flakiness is timing dependent.
+      await Future<void>.delayed(const Duration(seconds: 10));
+    }
+    rmTree(parent);
+  }
+
+}
+
+class _FlutterPlugin extends _FlutterProject {
+  _FlutterPlugin(Directory parent, String name) : super(parent, name);
+
+  static Future<_FlutterPlugin> create(
+      Directory directory, List<String> options) async {
     await inDirectory(directory, () async {
       await flutter(
         'create',
-        options: <String>['--template=app', '--org', 'io.flutter.devicelab', ...options, 'plugintest'],
+        options: <String>[
+          '--template=plugin',
+          '--org',
+          'io.flutter.devicelab',
+          ...options,
+          'plugintest'
+        ],
       );
     });
-    return FlutterProject(directory, 'plugintest');
+    return _FlutterPlugin(directory, 'plugintest');
   }
 
-  String get rootPath => path.join(parent.path, name);
+  Future<void> test() async {
+    await inDirectory(Directory(rootPath), () async {
+      await flutter('test');
+    });
+  }
+}
+
+class _FlutterApp extends _FlutterProject {
+  _FlutterApp(Directory parent, String name) : super(parent, name);
+
+  static Future<_FlutterApp> create(
+      Directory directory, List<String> options) async {
+    await inDirectory(directory, () async {
+      await flutter(
+        'create',
+        options: <String>[
+          '--template=app',
+          '--org',
+          'io.flutter.devicelab',
+          ...options,
+          'plugintestapp'
+        ],
+      );
+    });
+    return _FlutterApp(directory, 'plugintestapp');
+  }
 
-  Future<void> addPlugin(String plugin) async {
+  Future<void> addPlugin(String plugin, {String pluginPath}) async {
     final File pubspec = File(path.join(rootPath, 'pubspec.yaml'));
```

Comment:
Does `addPlugin` have to be a method on the _FlutterPlugin class? It looks to me the two classes `_FlutterProject` and `_FlutterPlugin` are quite similar except the binary/project name? Wondering whether you'll have to define two classes.

Reply:
`_FlutterPlugin` extends `_FlutterProject`.

I've moved `addPlugin` from `_FlutterApp` into `_FlutterProject` and `test` from `_FlutterPlugin` into `_FlutterProject` since they make sense for both plugins and apps. 

I haven't written any test that have plugins depending on plugins, but I added a line that ensures that the template app's unit tests pass.